src/virtual_machine/trap.rs

```diff
@@ -28,71 +28,71 @@ impl From<u16> for Trap {
+            Trap::PUTS => puts(virtual_machine),
+            Trap::PUTSP => putsp(virtual_machine),
+        }
     }
+}
 
-    fn halt(&self) {
-        std::process::exit(0);
-    }
+fn getc(virtual_machine: &mut LC3VirtualMachine) {
+    let mut buffer = [0; 1];
+    std::io::stdin()
+        .read_exact(&mut buffer)
+        .expect("Couldn't read from stdin");
+    virtual_machine.update_register(Register::R0, buffer[0] as u16);
+}
 
-    fn in_trap(&self, virtual_machine: &mut LC3VirtualMachine) {
-        println!("Enter a character: ");
-        let char = std::io::stdin()
-            .bytes()
-            .next()
-            .and_then(|read_result| read_result.ok())
-            .map(|char| char as u16)
-            .expect("Couldn't read from stdin");
-        virtual_machine.update_register(Register::R0, char);
-    }
+fn halt() {
+    std::process::exit(0);
+}
 
-    fn out(&self, virtual_machine: &mut LC3VirtualMachine) {
-        print!(
-            "{}",
-            (virtual_machine.read_register(Register::R0) as u8) as char
-        );
-    }
+fn in_trap(virtual_machine: &mut LC3VirtualMachine) {
+    println!("Enter a character: ");
+    let char = std::io::stdin()
+        .bytes()
+        .next()
+        .and_then(|read_result| read_result.ok())
+        .map(|char| char as u16)
+        .expect("Couldn't read from stdin");
+    virtual_machine.update_register(Register::R0, char);
+}
 
-    fn puts(&self, virtual_machine: &mut LC3VirtualMachine) {
-        let mut read_index = virtual_machine.read_register(Register::R0);
-        let mut char = virtual_machine.memory_read(read_index);
-        while char != 0 {
-            print!("{}", (char as u8) as char);
-            read_index += 1;
-            char = virtual_machine.memory_read(read_index);
-        }
-        std::io::stdout().flush().expect("Couldn't flush");
-    }
+fn out(virtual_machine: &mut LC3VirtualMachine) {
+    print!(
+        "{}",
+        (virtual_machine.read_register(Register::R0) as u8) as char
+    );
+}
 
-    fn putsp(&self, virtual_machine: &mut LC3VirtualMachine) {
-        let mut read_index = virtual_machine.read_register(Register::R0);
-        let mut char = virtual_machine.memory_read(read_index);
-        while char != 0 {
-            let first_char = char & 0b11111111;
-            print!("{}", (first_char as u8) as char);
-            let second_char = char >> 8;
-            if second_char != 0 {
-                print!("{}", (second_char as u8) as char);
-            }
-            read_index += 1;
-            char = virtual_machine.memory_read(read_index);
-        }
-        std::io::stdout().flush().expect("Couldn't flush");
+fn puts(virtual_machine: &mut LC3VirtualMachine) {
+    let mut read_index = virtual_machine.read_register(Register::R0);
+    let mut char = virtual_machine.memory_read(read_index);
+    while char != 0 {
+        print!("{}", (char as u8) as char);
+        read_index += 1;
```

Comment:
```suggestion
        read_index = read_index.wrapping_add(1);
```
To prevent overflow.